compileall2.py

```diff
@@ -256,14 +264,31 @@ def compile_file(fullname, ddir=None, force=False, rx=None, quiet=0,
             if not quiet:
                 print('Compiling {!r}...'.format(fullname))
             try:
-                for opt_level, cfile in opt_cfiles.items():
+                for index, opt_level in enumerate(sorted(optimize)):
+                    cfile = opt_cfiles[opt_level]
                     if PY37:
                         ok = py_compile.compile(fullname, cfile, dfile, True,
                                                 optimize=opt_level,
                                                 invalidation_mode=invalidation_mode)
                     else:
                         ok = py_compile.compile(fullname, cfile, dfile, True,
                                                 optimize=opt_level)
+                    
+                    if index > 0 and hardlink_dupes:
+                        previous_cfile = opt_cfiles[optimize[index - 1]]
+                        if previous_cfile == cfile and optimize[0] not in (1, 2):
+                            # Python 3.4 has only one .pyo file for -O and -OO so
+                            # we hardlink it only if there is a .pyc file
+                            # with the same content
+                            previous_cfile = opt_cfiles[optimize[0]]
+                        try:
+                            if  previous_cfile != cfile and filecmp.cmp(cfile, previous_cfile, shallow=False):
+                                os.unlink(cfile)
+                                os.link(previous_cfile, cfile)
+                        except OSError:
+                            # We cannot use hardlink deduplication
```

Comment:
one thought here: if the `OSError` is triggered during `os.link` we end up with the file deleted and missing.

Reply:
Well, if we are able to manipulate the file on a fs and the unlink step is okay, how likely the second step breaks? I mean, we can create a copy of `cfile` somewhere and copy it back if one of these two steps fail but the copying somewhere and back might fail as well which leads to neverending paranoic exception handling.

Do you have some specific corner case in mind which would makes sense to handle?

Reply:
Operating systems that don't support hardlinks will proceed with unlink just fine.

I don't think we need to do "move aside, try hardlink, delete" magic. We should simply fail here and say: It seems your system does not support hardlinking.

Reply:
In the end, the situation is a little bit different. The fact that Windows does not support hard links means that `os.link()` creates a usual copy of a file.
This means that we can remove try/except OSError because hard link deduplication basically does nothing on windows.

What do you think?

Reply:
Oh. Ok, get rid of it :tada: